Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in Source/WebCore/platform/graphics

cdumez opened the pull request:
#### f0f93c393abec436e0da6b4017dfd04f69f8f65f
<pre>
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in Source/WebCore/platform/graphics
<a href="https://bugs.webkit.org/show_bug.cgi?id=285367">https://bugs.webkit.org/show_bug.cgi?id=285367</a>

Reviewed by Darin Adler.

* Source/WebCore/platform/graphics/mac/ColorMac.mm:
* Source/WebCore/platform/graphics/mac/controls/ButtonMac.mm:
(WebCore::ButtonMac::cellSize const):
(WebCore::ButtonMac::cellOutsets const):
* Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.mm:
(WebCore::InnerSpinButtonMac::cellSize const):
* Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.mm:
(WebCore::interpolateGradient):
(WebCore::topGradientInterpolate):
(WebCore::bottomGradientInterpolate):
(WebCore::mainGradientInterpolate):
(WebCore::darkTopGradientInterpolate):
(WebCore::darkBottomGradientInterpolate):
(WebCore::darkMainGradientInterpolate):
* Source/WebCore/platform/graphics/mac/controls/MenuListMac.mm:
(WebCore::MenuListMac::cellSize const):
(WebCore::MenuListMac::cellOutsets const):
* Source/WebCore/platform/graphics/mac/controls/ProgressBarMac.mm:
(WebCore::ProgressBarMac::cellOutsets const):
* Source/WebCore/platform/graphics/mac/controls/SearchFieldCancelButtonMac.mm:
(WebCore::SearchFieldCancelButtonMac::cellSize const):
* Source/WebCore/platform/graphics/mac/controls/SliderTrackMac.mm:
(WebCore::trackGradientInterpolate):
* Source/WebCore/platform/graphics/mac/controls/SwitchMacUtilities.mm:
(WebCore::SwitchMacUtilities::cellOutsets):
* Source/WebCore/platform/graphics/mac/controls/ToggleButtonMac.mm:
(WebCore::ToggleButtonMac::cellOutsets const):
* Source/WebCore/platform/graphics/opentype/OpenTypeCG.cpp:
(WebCore::OpenType::readShortFromTable):
(WebCore::OpenType::tryGetTypoMetrics):
* Source/WebCore/platform/graphics/opentype/OpenTypeMathData.cpp:
(WebCore::OpenTypeMathData::OpenTypeMathData):
* Source/WebCore/platform/graphics/opentype/OpenTypeTypes.h:
(WebCore::OpenType::validateTableSingle):
(WebCore::OpenType::validateTable):
* Source/WebCore/platform/graphics/opentype/OpenTypeVerticalData.cpp:
(WebCore::loadHmtxTable):
(WebCore::OpenTypeVerticalData::loadMetrics):
(WebCore::OpenTypeVerticalData::loadVerticalGlyphSubstitutions):
* Source/WebCore/platform/graphics/transforms/TransformOperation.cpp:
(WebCore::TransformOperation::sharedPrimitiveType const):
* Source/WebCore/platform/graphics/transforms/TransformationMatrix.cpp:
(WebCore::decompose4):

Canonical link: <a href="https://commits.webkit.org/288440@main">https://commits.webkit.org/288440@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/503c242a39a03731b52959fce6dbf0c80e17976a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83345 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2964 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37636 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88424 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/34356 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/85436 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3047 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10910 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64839 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22585 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86398 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2216 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75746 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45123 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2123 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29947 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33407 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73238 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30674 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89798 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10612 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/7640 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73267 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10837 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71567 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72497 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16710 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15447 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/1947 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12870 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10567 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16039 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10417 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13888 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12190 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.mm:
```diff
@@ -45,53 +43,55 @@
-    static constexpr float dark[4] = { 1.0f, 1.0f, 1.0f, 0.4f };
-    static constexpr float light[4] = { 1.0f, 1.0f, 1.0f, 0.15f };
+    static constexpr std::array<float, 4> dark { 1.0f, 1.0f, 1.0f, 0.4f };
+    static constexpr std::array<float, 4> light { 1.0f, 1.0f, 1.0f, 0.15f };
```
```suggestion
    static constexpr std::array dark { 1.0f, 1.0f, 1.0f, 0.4f };
    static constexpr std::array light { 1.0f, 1.0f, 1.0f, 0.15f };
```
And similar in the rest of the file.

darinadler commented on Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.mm:
```diff
@@ -45,53 +43,55 @@
 {
     float a = inData[0];
     int i = 0;
     for (i = 0; i < 4; i++)
```
This is strange. Should just define `i` inside the `for` statement.

darinadler commented on Source/WebCore/platform/graphics/mac/controls/SliderTrackMac.mm:
```diff
@@ -64,15 +62,17 @@
-    static const float dark[4] = { 0.0f, 0.0f, 0.0f, 0.678f };
-    static const float light[4] = { 0.0f, 0.0f, 0.0f, 0.13f };
+    static constexpr std::array<float, 4> dark { 0.0f, 0.0f, 0.0f, 0.678f };
+    static constexpr std::array<float, 4> light { 0.0f, 0.0f, 0.0f, 0.13f };
```
```suggestion
    static constexpr std::array dark { 0.0f, 0.0f, 0.0f, 0.678f };
    static constexpr std::array light { 0.0f, 0.0f, 0.0f, 0.13f };
```

darinadler commented on Source/WebCore/platform/graphics/mac/controls/SliderTrackMac.mm:
```diff
@@ -64,15 +62,17 @@
+    static constexpr std::array<float, 4> light { 0.0f, 0.0f, 0.0f, 0.13f };
     float a = inData[0];
     int i = 0;
     for (i = 0; i < 4; i++)
```
Same strange `for` loop here, initializing twice.

darinadler commented on Source/WebCore/platform/graphics/opentype/OpenTypeTypes.h:
```diff
@@ -75,11 +74,19 @@ typedef UInt16 GlyphID;
-    if (!buffer || buffer->size() < sizeof(T) * count)
-        return 0;
-    return reinterpret_cast<const T*>(buffer->span().data());
+    size_t byteSize = sizeof(T) * count;
```
If count is high this can overflow unless we only pass constants in. We may be able to write this with integer division instead and have no risk of overflow.

darinadler commented on Source/WebCore/platform/graphics/opentype/OpenTypeVerticalData.cpp:
```diff
@@ -385,7 +381,7 @@ struct GSUBTable : TableBase {
 {
     auto buffer = platformData.openTypeTable(OpenType::HheaTag);
-    const OpenType::HheaTable* hhea = OpenType::validateTable<OpenType::HheaTable>(buffer);
+    const OpenType::HheaTable* hhea = OpenType::validateTableSingle<OpenType::HheaTable>(buffer);
```
```suggestion
    auto hhea = OpenType::validateTableSingle<OpenType::HheaTable>(buffer);
```

darinadler commented on Source/WebCore/platform/graphics/opentype/OpenTypeVerticalData.cpp:
```diff
@@ -428,7 +425,7 @@ void OpenTypeVerticalData::loadMetrics(const FontPlatformData& platformData)
     // Load vhea first. This table is required for fonts that support vertical flow.
     auto buffer = platformData.openTypeTable(OpenType::VheaTag);
-    const OpenType::VheaTable* vhea = OpenType::validateTable<OpenType::VheaTable>(buffer);
+    const OpenType::VheaTable* vhea = OpenType::validateTableSingle<OpenType::VheaTable>(buffer);
```
```suggestion
    auto vhea = OpenType::validateTableSingle<OpenType::VheaTable>(buffer);
```

darinadler commented on Source/WebCore/platform/graphics/opentype/OpenTypeVerticalData.cpp:
```diff
@@ -439,7 +436,7 @@ void OpenTypeVerticalData::loadMetrics(const FontPlatformData& platformData)
     // Load VORG. This table is optional.
     buffer = platformData.openTypeTable(OpenType::VORGTag);
-    const OpenType::VORGTable* vorg = OpenType::validateTable<OpenType::VORGTable>(buffer);
+    const OpenType::VORGTable* vorg = OpenType::validateTableSingle<OpenType::VORGTable>(buffer);
```
```suggestion
    auto vorg = OpenType::validateTableSingle<OpenType::VORGTable>(buffer);
```

darinadler commented on Source/WebCore/platform/graphics/opentype/OpenTypeVerticalData.cpp:
```diff
@@ -490,7 +487,7 @@ void OpenTypeVerticalData::loadMetrics(const FontPlatformData& platformData)
 {
     auto buffer = platformData.openTypeTable(OpenType::GSUBTag);
-    const OpenType::GSUBTable* gsub = OpenType::validateTable<OpenType::GSUBTable>(buffer);
+    const OpenType::GSUBTable* gsub = OpenType::validateTableSingle<OpenType::GSUBTable>(buffer);
```
```suggestion
    auto gsub = OpenType::validateTableSingle<OpenType::GSUBTable>(buffer);
```

cdumez committed:
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in Source/WebCore/platform/graphics
https://bugs.webkit.org/show_bug.cgi?id=285367

Reviewed by Darin Adler.

* Source/WebCore/platform/graphics/mac/ColorMac.mm:
* Source/WebCore/platform/graphics/mac/controls/ButtonMac.mm:
(WebCore::ButtonMac::cellSize const):
(WebCore::ButtonMac::cellOutsets const):
* Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.mm:
(WebCore::InnerSpinButtonMac::cellSize const):
* Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.mm:
(WebCore::interpolateGradient):
(WebCore::topGradientInterpolate):
(WebCore::bottomGradientInterpolate):
(WebCore::mainGradientInterpolate):
(WebCore::darkTopGradientInterpolate):
(WebCore::darkBottomGradientInterpolate):
(WebCore::darkMainGradientInterpolate):
* Source/WebCore/platform/graphics/mac/controls/MenuListMac.mm:
(WebCore::MenuListMac::cellSize const):
(WebCore::MenuListMac::cellOutsets const):
* Source/WebCore/platform/graphics/mac/controls/ProgressBarMac.mm:
(WebCore::ProgressBarMac::cellOutsets const):
* Source/WebCore/platform/graphics/mac/controls/SearchFieldCancelButtonMac.mm:
(WebCore::SearchFieldCancelButtonMac::cellSize const):
* Source/WebCore/platform/graphics/mac/controls/SliderTrackMac.mm:
(WebCore::trackGradientInterpolate):
* Source/WebCore/platform/graphics/mac/controls/SwitchMacUtilities.mm:
(WebCore::SwitchMacUtilities::cellOutsets):
* Source/WebCore/platform/graphics/mac/controls/ToggleButtonMac.mm:
(WebCore::ToggleButtonMac::cellOutsets const):
* Source/WebCore/platform/graphics/opentype/OpenTypeCG.cpp:
(WebCore::OpenType::readShortFromTable):
(WebCore::OpenType::tryGetTypoMetrics):
* Source/WebCore/platform/graphics/opentype/OpenTypeMathData.cpp:
(WebCore::OpenTypeMathData::OpenTypeMathData):
* Source/WebCore/platform/graphics/opentype/OpenTypeTypes.h:
(WebCore::OpenType::validateTableSingle):
(WebCore::OpenType::validateTable):
* Source/WebCore/platform/graphics/opentype/OpenTypeVerticalData.cpp:
(WebCore::loadHmtxTable):
(WebCore::OpenTypeVerticalData::loadMetrics):
(WebCore::OpenTypeVerticalData::loadVerticalGlyphSubstitutions):
* Source/WebCore/platform/graphics/transforms/TransformOperation.cpp:
(WebCore::TransformOperation::sharedPrimitiveType const):
* Source/WebCore/platform/graphics/transforms/TransformationMatrix.cpp:
(WebCore::decompose4):

Canonical link: https://commits.webkit.org/288440@main
